python/casmsmartphase/MNVMerge.py

```diff
@@ -0,0 +1,313 @@
+#   LICENSE
+# Copyright (c) 2018-2019 Genome Research Ltd.
+# Author: Cancer Genome Project cgphelp@sanger.ac.uk
+#
+#
+# This file is part of vcf_flag_modifier.
+#
+# vcf_flag_modifier is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Affero General Public License as published by the Free
+# Software Foundation; either version 3 of the License, or (at your option) any
+# later version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+# details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+#    1. The usage of a range of years within a copyright statement contained within
+#    this distribution should be interpreted as being equivalent to a list of years
+#    including the first and last year specified and all consecutive years between
+#    them. For example, a copyright statement that reads ‘Copyright (c) 2005, 2007-
+#    2009, 2011-2012’ should be interpreted as being identical to a statement that
+#    reads ‘Copyright (c) 2005, 2007, 2008, 2009, 2011, 2012’ and a copyright
+#    statement that reads ‘Copyright (c) 2005-2012’ should be interpreted as being
+#    identical to a statement that reads ‘Copyright (c) 2005, 2006, 2007, 2008,
+#    2009, 2010, 2011, 2012’."
+#
+#
+"""
+Python module for methods reading/parsing and merging MNVs in an
+unflagged (not post processed) CaVEMan generated VCF
+"""
+import datetime
+import os
+import re
+import sys
+
+import vcfpy
+
+# Setup base variables for the VCF process line
+base_vcf_process_key = "vcfProcessLog"
+base_vcf_process_log = "<InputVCF=<{}>,InputVCFSource=<{}>,InputVCFParam=<{}>>"
+
+
+def get_last_vcf_process_index(header_lines, key_prefix):
+    """
+    Find all header lines matching the prefix key_prefix.
+    Where they do match, find the greatest index, so we can increment
+    when generating a new header line for this process.
+    Returns greatest index found in the header key.
+    """
+    matching_head_keys = set(
+        line.key
+        for line in (
+            line
+            for line in header_lines
+            if isinstance(line, vcfpy.header.HeaderLine)
+            and line.key.startswith(key_prefix)
+        )
+    )
+    # Now find the maximum indices of the keys found
+    indices = [
+        int(key_part[1])
+        for key_part in (key.rsplit(".", 1) for key in matching_head_keys)
+        if len(key_part) == 2 and key_part[1].isdigit
+    ]
+    if indices:
+        return max(indices)
```

Comment:
Would be more readable to do a simple loop that updates `max_index` if its greater than create a list to then only look for the max value.  This is fine if you intend to use the list for something else, but not convinced it's necessary here.

You don't have to use comprehension for every loop action and if you have to start nesting them where a single loop could achieve the same thing it's probably not the best way.

Reply:
Is black capable of this level of reformatting? I don't recall writing this code. It makes my head hurt!